### ISSUE TITLE
👀 Upgrading from V8 to V9 causes transactions with custom names to disappear in UI and errors in the logs

### ISSUE THREAD

**Issue by NielsKSchjoedt** [newrelic/newrelic-ruby-agent#1859]
I'm in the process of upgrading a large Rails 6.2 app running APM 8.16 to version 9.0.

In the transactions interface on some web transactions (seems to be the ones with custom transaction names) I see now that most time is attributed to a segment by the name `#`. What is this and why?

![Skærmbillede 2023-03-06 kl  15 32 57](https://user-images.githubusercontent.com/570755/223373406-a697e27c-7300-4a4f-a468-62c00a81307c.png)

In the `newrelic.log` I can see the following:

```
[2023-03-06 15:13:18 +0100 web-01 (1107039)] ERROR : Exception during Transaction#finish
[2023-03-06 15:13:18 +0100 web-01 (1107039)] ERROR : NoMethodError: undefined method `name' for nil:NilClass
[2023-03-06 15:13:18 +0100 web-01 (1107039)] DEBUG : Debugging backtrace:
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction/abstract_segment.rb:267:in `force_finish'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction/abstract_segment.rb:91:in `finalize'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction/tracing.rb:52:in `block in finalize_segments'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction/tracing.rb:52:in `each'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction/tracing.rb:52:in `finalize_segments'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction.rb:555:in `commit!'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/transaction.rb:534:in `finish'
/home/app/a/shared/bundle/ruby/2.7.0/gems/newrelic_rpm-9.0.0/lib/new_relic/agent/instrumentation/middleware_tracing.rb:111:in `call'
/home/app/a/shared/bundle/ruby/2.7.0/gems/unicorn-5.8.0/lib/unicorn/http_server.rb:634:in `process_client'
/home/app/a/shared/bundle/ruby/2.7.0/gems/unicorn-5.8.0/lib/unicorn/http_server.rb:732:in `worker_loop'
/home/app/a/shared/bundle/ruby/2.7.0/gems/unicorn-5.8.0/lib/unicorn/http_server.rb:548:in `spawn_missing_workers'
/home/app/a/shared/bundle/ruby/2.7.0/gems/unicorn-5.8.0/lib/unicorn/http_server.rb:144:in `start'
/home/app/a/shared/bundle/ruby/2.7.0/gems/unicorn-5.8.0/bin/unicorn:128:in `&lt;top (required)&gt;'
/home/app/a/shared/bundle/ruby/2.7.0/bin/unicorn:23:in `load'
/home/app/a/shared/bundle/ruby/2.7.0/bin/unicorn:23:in `&lt;top (required)&gt;'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/cli/exec.rb:63:in `load'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/cli/exec.rb:63:in `kernel_load'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/cli/exec.rb:28:in `run'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/cli.rb:481:in `exec'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/cli.rb:31:in `dispatch'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/cli.rb:25:in `start'
/home/app/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.25/exe/bundle:49:in `block in &lt;top (required)&gt;'
/home/app/.rbenv/versions/2.7.2/lib/ruby/site_ruby/2.7.0/bundler/friendly_errors.rb:128:in `with_friendly_errors'
/home/app/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.25/exe/bundle:37:in `&lt;top (required)&gt;'
/home/app/.rbenv/versions/2.7.2/bin/bundle:23:in `load'
/home/app/.rbenv/versions/2.7.2/bin/bundle:23:in `&lt;main&gt;'
```

The custom naming in rails controller looks like this: 

```ruby
NewRelic::Agent.set_transaction_name("#{NewRelic::Agent.get_transaction_name} - search_backend: #{resolver.backend}")
```

I can see here that it seems to be `parent` which is nil https://github.com/newrelic/newrelic-ruby-agent/blob/68a34050cff2034559ea1acc59132d3ee245ce42/lib/new_relic/agent/transaction/abstract_segment.rb#L267 - How can I further debug this and what might be causing it and could it be related?

**Comment by NielsKSchjoedt** [comment]
This is how it looks in V8
![Skærmbillede 2023-03-07 kl  10 03 53](https://user-images.githubusercontent.com/570755/223374737-1cffa5b3-b43a-44b4-be35-cc9af1ebe413.png)

So it looks like it's the postgresql part missing actually

**Comment by kaylareopelle** [comment]
Hi @NielsKSchjoedt, one more thing -- `newrelic_agent.log` and `newrelic_audit.log` can contain sensitive information. 

🚨 _**Please do not post these files directly to GitHub.**_ 🚨 

If you choose to post any code snippets, please make sure sensitive information is redacted.

**Comment by boomer196** [comment]
Experiencing the same issue with 9.0.0.  We are coming from 8.16.0.  

https://forum.newrelic.com/s/hubtopic/aAX8W0000008eoaWAA/ruby-agent-has-nomethoderror-during-transactionfinish

**Comment by kaylareopelle** [comment]
Thanks @boomer196 for linking the two issues! 

We think the problem might be with our instrumentation for [Fibers](https://ruby-doc.org/3.2.1/Fiber.html) introduced in the 9.0 release.

Could you try disabling the Fiber instrumentation and let us know if the problem persists? 

To disable the instrumentation in your `newrelic.yml`, add: 
```yml
instrumentation.fiber: disabled 
```

To disable the instrumentation using an environment variable add: 
`NEW_RELIC_INSTRUMENTATION_FIBER=disabled`

cc @NielsKSchjoedt

**Comment by boomer196** [comment]
> Could you try disabling the Fiber instrumentation and let us know if the problem persists?
> 
> To disable the instrumentation in your `newrelic.yml`, add:
> 
> ```yaml
> instrumentation.fiber: disabled 
> ```
@kaylareopelle, I disabled the fiber instrumentation as you suggested and the problem still occurs for us.  As a side note do not use any custom names.  Our implementation is basically out of the box defaults.

**Comment by kaylareopelle** [comment]
Thanks for the follow-up, @boomer196! I'll keep searching for a solution. 

It sounds like you're using a Rails app. Which version are you on? Is your Ruby version 3.2? (it seems like this from the logs in the Explorer's Hub post)

**Comment by boomer196** [comment]
> Thanks for the follow-up, @boomer196! I'll keep searching for a solution.
> 
> It sounds like you're using a Rails app. Which version are you on? Is your Ruby version 3.2? (it seems like this from the logs in the Explorer's Hub post)

Yes @kaylareopelle, we are running `rails@7.0.4.2` with `ruby@3.2.1`

**Comment by kaylareopelle** [comment]
I have another config I'd like you to try, @boomer196. Could you turn off thread tracing?

```
instrumentation.thread.tracing: false 
```

Turning this off in a sandbox environment with a similar error seemed to stop the problem.

**Comment by boomer196** [comment]
@kaylareopelle, setting `instrumentation.thread.tracing: false` appears to have stopped the problem for us as well.

**Comment by kaylareopelle** [comment]
Thank you, @boomer196. This helps us narrow down which PRs may have introduced the bug! I'm going to do a bit more tinkering and will reach out with a bug fix branch when ready.

**Comment by kaylareopelle** [comment]
Hi @NielsKSchjoedt, @boomer196 - I have a bug fix branch I'd like you both to test out. It resolves the raised error, but I'm not sure if it'll solve the segments named `#`. 

To install it, please replace your current installation of `newrelic_rpm` in your Gemfile with: 
```
gem 'newrelic_rpm', git: 'https://github.com/newrelic/newrelic-ruby-agent.git', branch: 'bugfix/transaction_finish_nil_parent'
```

We'd also like you to turn thread instrumentation back on while testing. To do this, you can either change the value of `instrumentation.thread.tracing` to `true` or remove the `instrumentation.thread.tracing` line from your config.

**Comment by kaylareopelle** [comment]
@NielsKSchjoedt, we looked further into the way segments are reporting with Threads and noticed the `#` segments weren't present when using the "New view" for transactions. Threads in Ruby have been introduced since the new view was rolled out, so the old view doesn't have the same support for these types of segments.

When running in a sandbox Rack application that generates one thread per request, this is what we see in the old transaction view: 
<img width="764" alt="Old Transaction View" src="https://user-images.githubusercontent.com/87386821/224801444-810ac377-5e9c-4e18-9804-b802b2ca8c63.png">

And here's what we see in the new transaction view: 
<img width="812" alt="New Transaction View Graph" src="https://user-images.githubusercontent.com/87386821/224801762-a5dd6555-fce6-40d9-99ca-b8edac3bc0ab.png">
<img width="818" alt="New Transaction View Table" src="https://user-images.githubusercontent.com/87386821/224801833-69ccabaa-2051-4693-b7bd-0b8b3896b7f7.png">

In the bottom left corner of the graph, you can see `Ruby/Thread` and the segment is called `Ruby/Thread` in the table.

Since thread instrumentation was turned on automatically in 9.0, this seems like the reason why upgrading changed the segment names. Thread instrumentation is optional. You can disable it, as mentioned above, by adding to your `newrelic.yml`:

```yml
# newrelic.yml
instrumentation.thread.tracing: false 
```

**Comment by boomer196** [comment]
> Hi @NielsKSchjoedt, @boomer196 - I have a bug fix branch I'd like you both to test out. It resolves the raised error, but I'm not sure if it'll solve the segments named `#`.

@kaylareopelle, it did solve the raised error, but not the warnings.

It appears to happen for all of our controller actions.  Here is an example of the logged WARN:

```** [NewRelic][2023-03-13 21:51:40 +0000 web.1 (57)] WARN : Segment: Controller/Rack/Titan::Application/call was unfinished at the end of transaction. Timing information for this segment's parent  in Controller/store/product_detail_pages/show may be inaccurate.```

- Do you have any idea why these warnings are being logged?

**Comment by kaylareopelle** [comment]
@boomer196 - These warnings are most likely logged because an asynchronous process was running when the transaction completed. In this case the segment didn't finish through the standard path, but through a forced finish. The forced finish code path logs this warning.

This is a bit of technical debt around our asynchronous segment timing we've discussed, but have not addressed because we were uncertain about what solution would be beneficial for our customers. 

I turned my notes on the team's prior discussions into this issue: https://github.com/newrelic/newrelic-ruby-agent/issues/1884 

What would you like to happen in this scenario?

**Comment by boomer196** [comment]
@kaylareopelle, thanks for opening that other issue.  I will move over to #1884 to discuss desired behavior.  As far as the exception we were seeing, it appears that by adding the `Safe Navigation Operator` that you did in your bugfix branch, you have fixed the exception.  I will bow out of this issue for now as I think we are going to just move up to `v9.0.0`, so we can keep current, but just disable `thread.tracing` until other issue is resolved.  Thanks for your help.

**Comment by fallwith** [comment]
The changes provided by #1876 have been released in v9.1.0 of the `newrelic_rpm` gem.
